feat!:  Added getImageLayer to flame_tiled

munsterlander opened the pull request:
# Description

This is a very simple add to flame_tile to allow getting image layers.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the
relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [X] The title of my PR starts with a [Conventional Commit] prefix (`fix:`, `feat:`, `docs:` etc).
- [X] I have read the [Contributor Guide] and followed the process outlined for submitting PRs.
- [X] I have updated/added tests for ALL new/updated/fixed functionality. 
- [X] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`. 
- [X] I have updated/added relevant examples in `examples`. 

## Breaking Change

Does your PR require Flame users to manually update their apps to accommodate your change?

- [X] Yes, this is a breaking change. (Indicate it in the [Conventional Commit] prefix with a `!`,
  e.g. `feat!:`, `fix!:`).
- [ ] No, this is *not* a breaking change.

ufrshubham reviewed (commented):
I was thinking should we make it a generic method

```dart
T getLayer<T extends Layer>(String name){
    final t = map.layers.firstWhere((layer) {
      return layer is T && layer.name == name;
    });
    return t as T;
}
```

munsterlander commented:
Which I concur with and if @spydon wants to implement this, then we can go in and change all the docs, examples, and tests.  It is a breaking change though, so I don't know if a phased approach is better, i.e. have all 5 methods co-exist and mark the specific implementations as deprecated (even though 3 would be new), have just 2 methods and mark the old as deprecated with the generic one there, or just break it and only have 1.

munsterlander commented:
Documenting implementation in case there is a question: `getLayer<ObjectLayer>(”myObjectLayer");`

st-pasha reviewed (commented):
1. Looks like the three functions can be combined into a single template function.
2. Please use doc-comments (`///`) for documentation. You may read more about them in [this guide](https://dart.dev/guides/language/effective-dart/documentation#doc-comments).
3. Consider what would happen when the map doesn't have the requested layer.
4. Perhaps add a few tests?

munsterlander commented:
@st-pasha Yes, ideally it can return null to avoid errors with calls to non-existent layers. Ok, I will update this pr with the relevant change. Do we need to leave the existing method and just label it as deprecated?

st-pasha commented:
> Do we need to leave the existing method and just label it as deprecated?

yes, breaking changes should be avoided if possible. So, leaving the existing method as-is and marking it as `@Deprecated()` is the right idea.

munsterlander commented:
Ok, I think that does it.  We added the code, tests, examples, doc block, and docs.  I updated the PR to indicate blocking and remove unnecessary comments.

st-pasha commented on packages/flame_tiled/lib/src/renderable_tile_map.dart:
```diff
@@ -202,10 +202,27 @@ class RenderableTiledMap {
+      foundLayer = map.layers.firstWhere((layer) {
+        return layer is T && layer.name == name;
+      }) as T;
+    } on StateError {
```
Usually, it is not recommended to use exceptions for anything other than exceptional situations. This is because creating an exception object is a fairly inefficient process.
So, instead of a try-catch this could be rewritten as a simple for-loop, which will be simpler code too.

ufrshubham replied:
This is now resolved with [70963eb](https://github.com/flame-engine/flame/pull/1405/commits/70963eb3db1c3d70dd59415b02cbcda0ce9c5eae)

spydon commented on packages/flame_tiled/lib/src/renderable_tile_map.dart:
```diff
@@ -202,10 +202,27 @@ class RenderableTiledMap {
+
+  /// Returns a layer of type [T] with given [name] from all the layers
+  /// of this map. If no such layer is found, null is returned.
+  T? getLayer<T extends Layer>(String name) {
+    T? foundLayer;
+
+    for (var i = 0; i < map.layers.length; ++i) {
+      final layer = map.layers.elementAt(i);
+      if (layer is T && layer.name == name) {
+        foundLayer = layer;
+        break;
+      }
+    }
+
+    return foundLayer;
+  }
```
```suggestion
  T? getLayer<T extends Layer>(String name) {
    T? foundLayer;

    for (final layer in map.layers.length {
      if (layer is T && layer.name == name) {
        foundLayer = layer;
        break;
      }
    }

    return foundLayer;
  }
```

ufrshubham replied:
Thanks @spydon! I wasn't even aware of `firstWhereOrNull` 😄
